Add TLE9104 output diagnostics sensor

benediktibk opened the pull request:
Implement the output diagnostics of the low side switch TLE9104 as a sensor.

henrikbrixandersen reviewed (commented):
I think this ought to be done using an MFD. The approach taken here is the same as before we had `drivers/mfd/`.

benediktibk commented:
> I think this ought to be done using an MFD. The approach taken here is the same as before we had `drivers/mfd/`.

Okay, I'm gonna focus on this PR with the changes and continue working on #67030 as soon as we have settled the basics in this one.

benediktibk committed:
drivers: gpio: split up driver for TLE9104 into a MFD

Split up the driver for the power train switch TLE9104 into a
MFD and GPIO.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

benediktibk committed:
drivers: sensor: implement output diagnostics sensor for TLE9104

Implement a sensor for the output diagnostics of the power train
switch TLE9104.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

benediktibk committed:
tests: drivers: build_all: sensor: add TLE9104

Add an instance of the output diagnostics sensor for the
power train switch TLE9104 to the build all tests.

Signed-off-by: Benedikt Schmidt <benedikt.schmidt@embedded-solutions.at>

aasinclair commented on drivers/mfd/mfd_tle9104.c:
```diff
@@ -0,0 +1,682 @@
+					  TLE9104_OFFDIAGCFG_DIAGFILTCFG_LENGTH, 0);
+		break;
+	default:
+		LOG_ERR("invalid value %i for diagnostic filter time",
```
These are calculated at compile time, so this error message isn't needed.
You could calculate the value at compile time using DT_ENUM_IDX, then you wouldn't need the switch case.

benediktibk replied:
To be able to use DT_ENUM_IDX I would have to turn the enums into strings, if I understood it correctly. But these values actually represent numbers. Therefore it feels weird to turn them into strings.

I removed the default cases anyway, as they definitely won't be hit (unless there is a pointer going wild ...)

aasinclair replied:
DT_ENUM_IDX works with ints, e.g:
https://github.com/zephyrproject-rtos/zephyr/blob/f0b975032a68c2da38c34f01b836edb5e9e62b12/dts/bindings/pinctrl/ambiq%2Capollo4-pinctrl.yaml#L95-L106

benediktibk replied:
done

aasinclair commented on dts/bindings/mfd/infineon,tle9104.yaml:
```diff
@@ -0,0 +1,72 @@
+      - 1000
+      - 500
+      - 200
+    description: "diagnostic filter time in us"
```
Default values need justification (e.g. `default matches the power-on reset value` or similar)

benediktibk replied:
done

aasinclair commented on include/zephyr/drivers/sensor.h:
```diff
@@ -186,6 +186,10 @@ enum sensor_channel {
 	SENSOR_CHAN_GAUGE_DESIRED_VOLTAGE,
 	/** Desired charging current in mA */
 	SENSOR_CHAN_GAUGE_DESIRED_CHARGING_CURRENT,
+	/** Open load detected, boolean with one bit per output */
```
These should be in a private header file

benediktibk replied:
I actually have another PR (https://github.com/zephyrproject-rtos/zephyr/pull/67030) lined up which uses the same sensor channels. Should it then still be private, but duplicated?

aasinclair replied:
Yes, I think they should be private but duplicated

benediktibk replied:
done

benediktibk commented:
Rebased onto main to resolve merge conflicts.

benediktibk commented:
Force pushed a change to set val2 as well to avoid random garbage.

benediktibk commented:
Changed the order of initialization, it is necessary to configure the diagnostics before OUT_EN is set.

benediktibk commented:
@MaureenHelm Could you please take a look at this PR?

benediktibk commented:
Rebased onto main to resolve merge conflicts.

benediktibk commented:
Rebased onto main to resolve merge conflicts.

benediktibk commented:
Rebased onto main to resolve merge conflicts.
